### DISCUSSION TITLE
Help establishing RootCA and IntermediateCA for Azure VPN

### DISCUSSION THREAD

**Discussion by BR41N14C** [strongswan/strongswan#1714] [Q&A]
Hello,
I have spent the entire day working on this issue and I can't seem to find an end to it.
I have a Root CA, and want to establish a Intermediate CA that will be responsible for signing user certificates to use for the Azure VPN.
I am able to get it to work with just Root CA and User Certificate, but whenever I try with Intermediate CA I can never get it to work.
Could someone guide me in the right way, I have tried scouting the web and nothing:

This is what I have tried:
**Generate the RootCA:**
ipsec pki --gen --outform pem > caKey.pem 
ipsec pki --self --in caKey.pem --dn "CN=VPN CA" --ca --outform pem > caCert.pem
 
**Create the IntermediateCA:**
ipsec pki --gen --outform pem > IntCAKey.pem
 
**Sign the IntermediateCA as CA + flag as keyCertSign:**
ipsec pki --pub --in IntCAKey.pem | ipsec pki --issue --cacert caCert.pem --cakey caKey.pem --dn "CN=IntCA" --flag keyCertSign --ca --outform pem > IntCACert.pem

**Generate user certificate:**
export PASSWORD="password"
export USERNAME=“username”
 
ipsec pki --gen --outform pem > "${USERNAME}Key.pem"
 
ipsec pki --pub --in "${USERNAME}Key.pem" | ipsec pki --issue --cacert ../IntCACert.pem --cakey ../IntCAKey.pem --dn "CN=${USERNAME}" --san "${USERNAME}" --flag clientAuth --outform pem > "${USERNAME}Cert.pem"
 
Now make it a .p12 bundle:
openssl pkcs12 -in "${USERNAME}Cert.pem" -inkey "${USERNAME}Key.pem" -certfile ../IntCACert.pem -export -out "${USERNAME}.p12" -password "pass:${PASSWORD}"

**Comment by BR41N14C** [comment]
Any guidance?

**Comment by tobiasbrunner** [comment]
> I am able to get it to work with just Root CA and User Certificate, but whenever I try with Intermediate CA I can never get it to work.

What exactly doesn't work? Please see #196.

**Reply by BR41N14C** [reply]
The VPN Connection.
It doesn't produce any logs, nor does it specify the error. Just that an unexpected error occurred.

A solution I found was to simply use the Root CA cert in the Azure Portal, and then use the Intermediate CA to sign off new User certificates. Perhaps, the entire problem was due to me entering the Intermediate CA cert into Azure instead of the Root CA.
However, to me that made sense... is the case that I NEED to add Root CA cert into Azure?

**Reply by tobiasbrunner** [reply]
Possible that it needs the complete certificate chain even if you explicitly add the intermediate CA certificate to the trust store. But you have to ask Microsoft about details.

**Reply by BR41N14C** [reply]
well it worked with just adding the Root CA in the trust store, I can sign off with the Intermediate CA no problem. I suppose that's the solution.
